File: Python/Searching/linear_search.py
'''
Approach Explanation:
- In linear search, we iterate through each element of the given array/list and compare it with the target element.
- If the target element is found, we return its index.
- If the entire array is traversed and the target element is not found, we return -1.

Time Complexity: 
- The time complexity of linear search is O(n), where n is the number of elements in the array. This is because, in the worst case scenario, we may have to traverse through all the elements in the array.

Space Complexity:
- The space complexity of linear search is O(1). This is because it only requires a constant amount of additional space to store the target element and loop variables.

Sample Input:
arr = [11, 45, 87, 23, 41, 92, 79]
target = 41


Sample Output:
Element 41 found at index 4
'''

#Here's a code for linear search in Python:

def linear_search(arr, target):
    """
    Function to perform a linear search in an array.
    
    Parameters:
    arr (list): List of elements
    target: Element to search
    
    Returns:
    int: Index of the target element if found, otherwise -1
    """
    
    # Traverse through each element of the array
    for i in range(len(arr)):
        # If the target element is found, return its index
        if arr[i] == target:
            return i
    
    # If the target element is not found, return -1
    return -1

File: C++/Searching/LinearSearch.cpp
/*
Explanation:
- The `linearSearch` function takes an array, its size, and a key as input. It performs a linear search by iterating through each element of the array using a `for` loop.
- In each iteration, it checks if the current element matches the key. If the key is found, it returns the index of that element.
- If the key is not found after scanning the entire array, the function returns -1 to indicate that the key is not present in the array.
- The main function contains a sample input array, a key to search for, calls the `linearSearch` function, and prints the result.

Sample Input:
Array: {10, 5, 3, 7, 2, 8}
Key: 7

Sample Output:
Key 7 found at index 3

Time Complexity: The time complexity of linear search is O(N), where N is the size of the array. In the worst case, the algorithm may need to traverse the entire array to find the key or determine that it is not present.
Space Complexity: The space complexity is O(1) since the algorithm only uses a constant amount of extra memory for the `i`, `size`, and `key` variables, regardless of the array size.
*/

#include <iostream>

// Function to perform linear search
int linearSearch(int arr[], int size, int key) {
    for (int i = 0; i < size; i++) {
        if (arr[i] == key) {
            // If key is found, return its index
            return i;
        }
    }

    // If key is not found, return -1
    return -1;
}

int main() {
    // Sample input
    int arr[] = {10, 5, 3, 7, 2, 8};
    int size = sizeof(arr) / sizeof(arr[0]);
    int key = 7;

    // Perform linear search to find the key in the array
    int index = linearSearch(arr, size, key);

    // Check if the key was found and print the result
    if (index != -1) {
        std::cout << "Key " << key << " found at index " << index << std::endl;
    } else {
        std::cout << "Key " << key << " not found in the array." << std::endl;
    }

    return 0;
}

